language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/codegen/PrimitiveValueDecoders.java

```diff
@@ -116,17 +117,67 @@ private static IllegalArgumentException mismatched(Class<?> clazz) {
    * @hidden
    */
   public static List<com.daml.ledger.javaapi.data.DamlRecord.Field> recordCheck(
-      int expectedFields, Value maybeRecord) {
+      int expectedFields, int optionalFields, Value maybeRecord) {
     var record =
         maybeRecord
             .asRecord()
             .orElseThrow(
                 () -> new IllegalArgumentException("Contracts must be constructed from Records"));
     var fields = record.getFields();
     var numberOfFields = fields.size();
-    if (numberOfFields != expectedFields)
-      throw new IllegalArgumentException(
-          "Expected " + expectedFields + " arguments, got " + numberOfFields);
+    if (numberOfFields == expectedFields) {
+      return fields;
+    }
+    if (numberOfFields > expectedFields) {
+      // Downgrade, check that the additional fields are empty optionals.
+      for (var i = expectedFields; i < numberOfFields; i++) {
+        final var field = fields.get(i);
+        final var optValue = field.getValue().asOptional();
+        if (optValue.isEmpty()) {
+          throw new IllegalArgumentException(
+              "Expected "
+                  + expectedFields
+                  + " arguments, got "
+                  + numberOfFields
+                  + " and field "
+                  + i
+                  + " is not optional: "
+                  + field);
+        }
+        final var value = optValue.get();
+        if (!value.isEmpty()) {
+          throw new IllegalArgumentException(
+              "Expected "
+                  + expectedFields
+                  + " arguments, got "
+                  + numberOfFields
+                  + " and field "
+                  + i
+                  + " is Optional but not empty: "
+                  + field);
+        }
+      }
```

Comment:
We should probably log upgrades and downgrades, perhaps at a low log level as there could be a lot of them.
```suggestion
      }
      logger.debug("Downgrading record, dropping {} trailing optional fields", (numberOfFields - expectedFields));
```
We could even be more descriptive by passing in the name of the record, and mentioning the names of the dropped fields.

Something like `LOG_EVERY_N` might be nice
https://github.com/google/glog#conditional-occasional-logging
but a low log level is probably enough.

Reply:
I've moved it to a trace log. I also had to add a dependency for slf4j as there is apparently nothing else in the java bindings that logs atm.